beets/util/__init__.py

```diff
@@ -438,51 +439,57 @@ def syspath(path: Bytes_or_String, prefix: bool = True) -> Bytes_or_String:
     if os.path.__name__ != "ntpath":
         return path
 
-    if not isinstance(path, str):
+    if isinstance(path, bytes):
         # Beets currently represents Windows paths internally with UTF-8
         # arbitrarily. But earlier versions used MBCS because it is
         # reported as the FS encoding by Windows. Try both.
         try:
-            path = path.decode("utf-8")
+            str_path = path.decode("utf-8")
         except UnicodeError:
             # The encoding should always be MBCS, Windows' broken
             # Unicode representation.
             assert isinstance(path, bytes)
             encoding = sys.getfilesystemencoding() or sys.getdefaultencoding()
-            path = path.decode(encoding, "replace")
+            str_path = path.decode(encoding, "replace")
+    else:
+        str_path = path
 
     # Add the magic prefix if it isn't already there.
     # https://msdn.microsoft.com/en-us/library/windows/desktop/aa365247.aspx
-    if prefix and not path.startswith(WINDOWS_MAGIC_PREFIX):
-        if path.startswith("\\\\"):
+    if prefix and not str_path.startswith(WINDOWS_MAGIC_PREFIX):
+        if str_path.startswith("\\\\"):
             # UNC path. Final path should look like \\?\UNC\...
-            path = "UNC" + path[1:]
-        path = WINDOWS_MAGIC_PREFIX + path
+            str_path = "UNC" + str_path[1:]
+        str_path = WINDOWS_MAGIC_PREFIX + str_path
 
-    return path
+    return str_path if isinstance(path, str) else str_path.encode()
 
 
-def samefile(p1: bytes, p2: bytes) -> bool:
+def samefile(p1: AnyStr, p2: AnyStr) -> bool:
     """Safer equality for paths."""
     if p1 == p2:
         return True
-    return shutil._samefile(syspath(p1), syspath(p2))
+    with suppress(OSError):
+        return os.path.samefile(syspath(p1), syspath(p2))
+
+    return False
```

Comment:
```suggestion
   return shutil._samefile(syspath(p1), syspath(p2))  # 2)) # type: ignore[attr-defined]
```

Did you check why we're using the internal `shutil` function here in the first place? Unless we properly understand this, let's not change the code, but instead ignore the typing error.

Reply:
I did - `samefile` was not supported on Windows until Python 3.2, see [here](https://docs.python.org/3/library/os.path.html#os.path.samefile)


> os.path.samefile(path1, path2)
Return True if both pathname arguments refer to the same file or directory. This is determined by the device number and i-node number and raises an exception if an [os.stat()](https://docs.python.org/3/library/os.html#os.stat) call on either pathname fails.
Changed in version 3.2: Added Windows support.
Changed in version 3.4: Windows now uses the same implementation as all other platforms.
Changed in version 3.6: Accepts a [path-like object](https://docs.python.org/3/glossary.html#term-path-like-object).

Reply:
Ok, and it appears that `shutil._samefile` also handles `OSError` in a similar way.

Any reason this is not just
```python
try:
    return os.path.samefile(...)
except OSError:
    return False
```

It's not more verbose in this case and the much more common idiom.

Reply:
Perplexity summarised what I wanted to respond with very well
![image](https://github.com/beetbox/beets/assets/16212750/f4522776-b03d-41c8-9a9a-c9dcfa8a54cc)

I initially learned about this from `flake8-simplify` [SIM-105](https://github.com/MartinThoma/flake8-simplify?tab=readme-ov-file#sim105)

```python
# Bad
try:
    foo()
except ValueError:
    pass

# Good
from contextlib import suppress

with suppress(ValueError):
    foo()
```

Reply:
I'd disagree in this case, because the `except` arm is not `pass`, but rather an actual statement. In any case, I don't have a very strong opinion here, go with the option that you prefer.